virtNode/projectQSimulator.py

```diff
@@ -0,0 +1,393 @@
+        return outcome
+
+    def measure_qubit(self, qubitNum):
+        """
+        Measures the desired qubit in the standard basis. This returns the classical outcome and deletes the qubit.
+
+        Arguments:
+        qubitNum	qubit to be measured
+        """
+        outcome = self.measure_qubit_inplace(qubitNum)
+
+        self.qubitReg.pop(qubitNum)
+
+        # Update the number of qubits
+        self.activeQubits = self.activeQubits - 1
+
+        return outcome
+
+    def replace_qubit(self, qubitNum, state):
+        """
+        Replaces the qubit at position qubitNum with the one given by state.
+        """
+        raise NotImplementedError("Currently you cannot replace a qubit using project Q as backend")
+
+    def absorb(self, other):
+        """
+        Absorb the qubits from the other engine into this one. This is done by tensoring the state at the end.
+        """
+
+        # Check whether there is space
+        newNum = self.activeQubits + other.activeQubits
+        if newNum > self.maxQubits:
+            raise quantumError("Cannot merge: qubits exceed the maximum available.\n")
+
+        # Check whether there are in fact qubits to tensor up....
+        if self.activeQubits == 0:
+            self.eng = other.eng
+            self.qubitReg = list(other.qubitReg)
+        elif other.activeQubits > 0:
+            # Get the current state of the other engine
+            other.eng.flush()
+            other_state = other.eng.backend.cheat()[1]
+
+            # Allocate qubits in this engine for the new qubits from the other engine
+            qreg = self.eng.allocate_qureg(other.activeQubits)
+
+            # Put the new qubits in the correct state
+            pQ.ops.StatePreparation(other_state) | qreg
+
+            # Add the qubits to the list of qubits
+            self.qubitReg += list(qreg)
+
+        self.activeQubits = newNum
+
+    def absorb_parts(self, R, I, activeQ):
+        """
+        Absorb the qubits, given in pieces
+
+        Arguments:
+        R		real part of the qubit state as a list
+        I		imaginary part as a list
+        activeQ		active number of qubits
+        """
+        # Check whether there is space
+        newNum = self.activeQubits + activeQ
+        if newNum > self.maxQubits:
+            raise quantumError("Cannot merge: qubits exceed the maximum available.\n")
+
+        if activeQ > 0:
+
+            # Convert the real and imaginary parts to a state
+            state = [re + im * 1j for re, im in zip(R, I)]
+
+            # Allocate qubits in this engine for the new qubits from the other engine
+            qreg = self.eng.allocate_qureg(activeQ)
+
+            # Put the new qubits in the correct state
+            pQ.ops.StatePreparation(state) | qreg
+
+            # Add the qubits to the list of qubits
+            self.qubitReg += list(qreg)
+
+            self.activeQubits = newNum
+
+
+class quantumRegister(projectQEngine):
```

Comment:
Same name as in the 'crudeSimulator'  
Do we need these classes, or can we just incorporate them in the super class?

(Also, classes should start with uppercase ;))

Reply:
This is the version of the classes which also stores the node etc. Do you suggest we change it?